designs/2018-processors-improvements/README.md

```diff
@@ -0,0 +1,241 @@
+    },
+    processor: "markdown/markdown"
+};
+```
+
+This configuration file will apply the named processor `markdown` from the `eslint-plugin-markdown` plugin to all files.
+
+The `processor` option would be interpreted inside of `CLIEngine` to determine which processors to apply to files before linting should begin.
+
+### Apply Processors Using `overrides`
+
+Of course, in most cases the end-user will not want to apply a processor to all files. In that case, the `overrides` configuration option may be used to narrowly define to which files the processors should be applied. For example:
+
+```js
+module.exports = {
+    plugins: ["markdown"],
+    parserOptions: {
+        ecmaVersion: 2018
+    },
+    overrides: [
+        {
+            files: ["*.md"],
+            processor: "markdown/markdown"
+        }
+    ]
+};
+```
+
+In this configuration, the named processor `markdown` from the `eslint-plugin-markdown` plugin is applied only to files with the extension `*.md`. This mimics the behavior of the extension-named processor that `eslint-plugin-markdown` currently exports.
+
+The `CLIEngine` method `getConfigForFile()` would automatically get the correct value for `processor` when the configuration is calculated.
+
+**Note:** End-users would still need to provide `--ext .md` on the command line for ESLint to automatically lint files ending with `.md`. No part of this proposal changes the behavior of `--ext`.
+
+### Virtual Filenames for Code Blocks
+
+The `preprocess()` method inside of a processor can now return an array containing either strings (the current behavior) or an object with two properties:
+
+* `filename` - (string) a virtual file filename to map the code block to (cannot contain `\` or `/` characters)
+* `text` - (string) the source code text
+
+For example:
+
+```js
+module.exports = {
+    preprocess(text, filename) {
+        
+        const { jsText1, jsText2 } = doSomething(text, filename);
+        
+        return [
+            {
+                filename: "1.js",
+                text: jsText1
+            },
+            {
+                filename: "2.js",
+                text: jsText2
+            }
+        ]
+    }
+};
+```
+
+This processor returns two code blocks containing JavaScript code. Each code block is given a virtual filename ending with `.js`.
+
+When a `preprocess()` method returns an object with a `filename` property, `CLIEngine` will call `getConfigForFile()` on the `filename` property to determine the correct configuration for the code block, which includes whether another processor should be run on the code block (matched by file extension).
+
+When a `preprocess()` method returns only a string, `CLIEngine` will interpret that as a JavaScript file and call `getConfigForFile()` using the parent file's filename appended with `".js"` (in case the parent filename does not end with `".js"`) to determine the correction configuration for the code block using this algorithm:
```

Comment:
I found a new concern:

Some `eslint-plugin-vue` users have their setting on `overrides` section:

```yml
#...
overrides:
  - files: "*.vue"
    parser: vue-eslint-parser
    plugins:
      - vue
    rules:
      vue/foo: error
      vue/bar: error
#...
```

And the plugin has an extension-named processor.

In this case, are the `vue/*` rules will get ignored because the code will get `.js` extension?
It looks a breaking change.

Reply:
Hmmm, I think you are right. This would break with the current proposal. What does `vue-eslint-parser` do?

Reply:
`vue-eslint-parser` parses `*.vue` files as HTML and generates [Vue.js Template AST](https://github.com/mysticatea/vue-eslint-parser/blob/59d52edb9d01989713f06c82618ce5c2b74a5646/docs/ast.md) from the first `<template>` element. The parser provides `parserServices` to access to the template AST. Also, it generates ESLint AST from the first `<script>` element and returns it.

On the other hand, the processor of `eslint-plugin-vue` provides the feature of `<!-- eslint-disable -->`-like directive comments. That `postprocess` method filters messages it should ignore. In fact, that `preprocess` method does nothing.

Reply:
Ah interesting. Thanks for explaining.

So is there a way we can do a non-breaking change to allow this behavior? (I don't think so, but I'm not 100% sure?)

If not, how would you be able to use `vue-eslint-parser` with this proposal?

Reply:
I think that we can keep the current behavior as is:

- it keeps the filename as is if `preprocess` returned a string.
- because the filename was not changed, it doesn't re-resolve the configuration and lints the code.

Reply:
Okay, just to be 100% sure, you're saying we can keep what's in the current proposal and it will not be a breaking change?

Reply:
@mysticatea can you answer my last question please?

Reply:
I'm sorry for overlooking.

I meant that we can keep the behavior when a processor's preprocess returned a string, as same as the current behavior. I.e., if the preprocess returned a string without a filename, ESLint doesn't change the filename that ESLint is verifying and call `Linter#verify`.

Reply:
Oh I see. And that will make the `vue-eslint-parser` use case work correctly? So this is NOT a breaking change?

Reply:
> And that will make the vue-eslint-parser use case work correctly? So this is NOT a breaking change?

Yes.

> an infinite loop

We can generalize it a bit. This will happen when a processor returned the same filename as parent's. My proposal (using the parent filename if a processor omitted a code block's filename) matches the condition.

My original proposal had a few lines to avoid infinity looping: https://github.com/eslint/rfcs/pull/1/files#diff-ae4914c864eeee75706376579948dcd1R152